dgtlmoon/changedetection.io
# Parse JSON documents embedded in HTML

This PR adresses issues discussed in #1495. 

## Problem

`driver.page_source` may sometimes follow the structure of:
```html
<html>
<head></head>
<body>
<pre style="...">
{}  # some json object
</pre>
</body>
</html>
```
There's no way of applying json filters on top of such a page. 

## Solution
The `extract_json_as_string` function currently checks all `<script></script>` tags to look for valid json object. The fix in this PR extends this functionality to also check whether the whole page body is a valid json object - if so, and the other approaches have failed, we will use it.

Note: these changes use `.text` instead of `.string` for the legacy behaviour (extracting json from <script> tags). This means that in case there were some html tags nested within <script> and they contained some text, it will be picked up. I figured this should not be an issue. If it is, though, let me know and I can use `.text` only for the `<body>` check. Apart from this and the new feature, the legacy behaviour should stay the same.

## Testing
I added one unit test for the new behaviour. The relevant tests are green. However, I wasn't able to run all of the tests for the project locally, since I haven't found instructions on how to do it in the repo. 

<details>
<summary>I also manually tested for the error mentioned in the original issue</summary>

1. Set up a watch with the following url: https://raw.githubusercontent.com/platformio/platform-espressif32/master/boards/lolin_s3.json
2. Select "Playwright Chromium/Javascript" in the "Request" 
3. Enter "json:$" to the "CSS/JSONPath/JQ/XPath Filters" in the "Filters & Triggers" 
4. Let it 
5. ~~See error "No parsable JSON found in this document"~~ The json should be extracted nicely as was the case for me:
<img width="618" alt="reproduced" src="https://github.com/dgtlmoon/changedetection.io/assets/15965147/766f2c31-897d-4bf6-be6d-e9418166476c">

</details>

--- 
`CONTRIBUTING.md` mentions I should merge into dev, but it seems like it's not an active branch and the recent PRs were against `master`.
